### ISSUE TITLE
Dream leaks sockets on macOS

### ISSUE THREAD

**Issue by polytypic** [aantron/dream#347]
On macOS [a simple Dream server](https://github.com/polytypic/dream-leak/blob/main/bin/server/server.ml),

```ocaml
let () =
  Dream.router
    [
      Dream.get "/" (fun _ ->
          Lwt.return
            (Dream.response
               ~headers:[ ("Content-Type", "text/plain") ]
               "On macOS I leak.\n"));
    ]
  |> Dream.logger |> Dream.run ~port:8080
```

leaks sockets (does not close them properly) and crashes after about 1k requests:

```
...
13.09.24 13:37:02.468    dream.logger  INFO REQ 1011 GET / ::1:50350 curl/8.7.1
On macOS I leak.
13.09.24 13:37:02.468    dream.logger  INFO REQ 1011 200 in 53 μs
13.09.24 13:37:02.481    dream.logger  INFO REQ 1012 GET / ::1:50351 curl/8.7.1
13.09.24 13:37:02.481    dream.logger  INFO REQ 1012 200 in 46 μs
On macOS I leak.
Fatal error: exception Unix.Unix_error(Unix.EINVAL, "select", "")
Raised by primitive operation at Lwt_engine.select#select in file "src/unix/lwt_engine.ml", line 407, characters 26-60
Called from Lwt_engine.select_based#iter in file "src/unix/lwt_engine.ml", line 348, characters 8-39
Called from Lwt_main.run.run_loop in file "src/unix/lwt_main.ml", line 36, characters 6-49
Called from Lwt_main.run in file "src/unix/lwt_main.ml", line 106, characters 8-13
Re-raised at Lwt_main.run in file "src/unix/lwt_main.ml", line 112, characters 4-13
Called from Dune__exe__Server in file "bin/server/server.ml", lines 2-10, characters 2-41
```

See [here](https://github.com/polytypic/dream-leak) for a full project that reproduces the issue with GitHub actions.

**Comment by aantron** [comment]
Thank you! This appears to be a very serious bug.

For confirmation, could you try with `alpha7` and confirm whether `conf-libev` is installed or not when you observe this issue, as I am concerned by the usage of `select`, as opposed to `kevent`?

**Comment by polytypic** [comment]
Yes, I have those locally and they also seem to installed to the CI machine.

```
...
  ∗ conf-libev            4-12         [required by dream]
...
  ∗ dream                 1.0.0~alpha7 [required by dream_leak]
...
```

**Comment by craff** [comment]
I think I have the same issue on a raspberry pi. I was using hyper, tried to close the message in Hyper.get / Hyper.post doest not solve the problem. 

Are you letting the GC close the socket ? This is unsafe, no warranty that the GC (especially the major GC) will collect socket fast enough ? 

I ask this, because socket are not freed immediatly after Hyper.get and Hyper.close exit. So my guess was socket closed by a finaliser in the GC.

**Comment by craff** [comment]
I copy the message I dit on hyper:

I investigated more. Calling Message.close and/or using Hyper.request did not solve the problem. 
I think it is deeper in Dream and may be related to [issue 347 of dream](https://github.com/aantron/dream/issues/347).

The observed behavior is that in general 5-10 unused sockets are preserved for some unknown reasons, but after a few days, this increase to more that 1000 sockets and the program fails with EMFILES as expected.

Switching to http-lwt-client immediatly solved the issue, so it must be an issue related to Hyper/Dream. Now no unused socket is kept.

Now that I have switched my code in production to http-lwt-client, I will not be able to investigate more. Sorry.

**Comment by aantron** [comment]
Thanks. I am still working on this, but greatly slowed down by that I haven't been able to observe it locally -- so far only in GHA.

**Comment by craff** [comment]
Anton Bachin ***@***.***> writes:

> Thanks. I am still working on this, but greatly slowed down by that I haven't been able to observe it locally -- so far only in GHA.

Try to just do a small client doing request in a loop, using
Hyper.get. You will observe that some sockets remain open. The behavior
is strange, and really looks like socket closed by a GC finaliser: there
are two many sockets open, but not one per request.

Cheers,
Christophe

> —
> Reply to this email directly, view it on GitHub, or unsubscribe.
> You are receiving this because you commented.


--
Christophe Raffalli
web: https://raffalli.eu
